File: python/tests/harness/vector_runner.py
"""Feeds one vector's events to a real Engine with a FakePlatform (plan Task 10, vector harness rules)."""

from __future__ import annotations

import json
from collections import deque
from typing import Any

import shared_files
from gibberish_rewriter.core.chord_detector import ChordAction
from gibberish_rewriter.core.config import AppConfig, Chord, ConfigParser
from gibberish_rewriter.core.engine import Engine, KeyEvent
from gibberish_rewriter.core.key_names import HOLD_KEY_ORDER, VK_PACKET, HoldKeys, KeyNames
from gibberish_rewriter.core.key_table import KeyTable, LayoutKind
from gibberish_rewriter.core.text_converter import TextConverter
from gibberish_rewriter.core.word_list import WordList

from harness.fake_platform import FakePlatform

NAMES = KeyNames.load(shared_files.path_of("keys.json"))
TABLE = KeyTable.load(shared_files.path_of("layouts/us-arabic101.json"))
CONVERTER = TextConverter(TABLE, WordList.load(shared_files.path_of("wordlists/english.txt")))
PARSER = ConfigParser(NAMES, shared_files.read_text("config.default.json"))

_HOLD_KEY_NAMES = {
    HoldKeys.SHIFT: "ShiftLeft",
    HoldKeys.CTRL: "ControlLeft",
    HoldKeys.ALT: "AltLeft",
    HoldKeys.WIN: "MetaLeft",
}


def load_vectors(relative: str) -> list[dict[str, Any]]:
    """Every vector of one file, in order. Names are unique within a file, as the C# loader also demands."""
    vectors: list[dict[str, Any]] = json.loads(shared_files.read_text(relative))
    names = [vector["name"] for vector in vectors]
    if len(set(names)) != len(names):
        raise ValueError(f"Vector names are not unique in {relative}.")
    return vectors


def vector_named(relative: str, name: str) -> dict[str, Any]:
    """The one vector of a file with this name."""
    return next(vector for vector in load_vectors(relative) if vector["name"] == name)


def _flag(vector: dict[str, Any], name: str) -> bool:
    """An optional boolean field, false when it is missing. Only a JSON boolean counts, as C# GetBoolean demands."""
    value = vector.get(name, False)
    if not isinstance(value, bool):
        raise ValueError(f"Vector field {name!r} must be a JSON boolean, not {json.dumps(value)}.")
    return value


def _call_value(value: Any) -> str:
    """One expected-call value, as the C# renders it: GetInt32 for a number, GetString for the rest, so a JSON
    null becomes an empty string and a boolean is an error."""
    if isinstance(value, str):
        return value
    if value is None:
        return ""
    if isinstance(value, int) and not isinstance(value, bool):
        return str(value)
    raise ValueError(f"Expected call value must be a string, a null or an integer, not {json.dumps(value)}.")


def _parse_layout(name: str) -> LayoutKind | None:
    return None if name == "other" else LayoutKind.parse(name)


def _hold_key_name(hold: HoldKeys) -> str:
    return _HOLD_KEY_NAMES.get(hold, "CapsLock")


class VectorRunner:
    """Feeds one vector's events to a real Engine with a FakePlatform (plan Task 10, vector harness rules)."""

    def __init__(self, vector: dict[str, Any]) -> None:
        self._down: set[int] = set()
        self._started: deque[ChordAction] = deque()
        self.decisions: list[str] = []
        """One decision string per key event."""

        config_text = json.dumps(vector["config"]) if "config" in vector else "{}"
        config = PARSER.parse(config_text).config
        if config is None:
            raise ValueError(f"Invalid config in vector: {config_text}")
        self.config: AppConfig = config

        self.platform = FakePlatform()
        self.platform.layout = _parse_layout(vector["layout"]) if "layout" in vector else LayoutKind.LATIN
        self.platform.is_console = _flag(vector, "console")
        self.platform.is_blocked_by_elevation = _flag(vector, "blockedByElevation")
        self.platform.release_in_time = not _flag(vector, "releaseTimeout")
        self.platform.clipboard_text = vector.get("clipboard")
        self.platform.selection = vector.get("selection")
        self.platform.clipboard_locked = vector.get("clipboardLocked")

        self.engine = Engine(
            NAMES,
            TABLE,
            CONVERTER,
            self.config,
            self.platform,
            self._started.append,
            _flag(vector, "capsLock"),
        )

    @staticmethod
    def expected_calls(vector: dict[str, Any]) -> list[str]:
        """Expected platform calls of a vector, as strings comparable with FakePlatform.calls."""
        calls = []
        for call in vector["expect"]:
            ((field, value),) = call.items()
            calls.append(f"{field} {_call_value(value)}")
        return calls

    def run(self, events: list[dict[str, Any]]) -> None:
        for event in events:
            self._feed(event)

    def _feed(self, event: dict[str, Any]) -> None:
        if "tap" in event:
            shift = _flag(event, "shift")
            if shift:
                self._key("ShiftLeft", True)
            self._key(event["tap"], True)
            self._key(event["tap"], False)
            if shift:
                self._key("ShiftLeft", False)
        elif "down" in event:
            self._key(event["down"], True)
        elif "up" in event:
            self._key(event["up"], False)
        elif "chord" in event:
            self._press_chord(event["chord"], event.get("during"))
        elif "mouseDown" in event:
            self.engine.on_mouse_down()
        elif "window" in event:
            self.platform.window = event["window"]
        elif "layout" in event:
            self.platform.layout = _parse_layout(event["layout"])
        elif "packet" in event:
            for _ in event["packet"]:
                self._send(VK_PACKET, True, is_packet=True)
                self._send(VK_PACKET, False, is_packet=True)
        elif "enabled" in event:
            self.engine.enabled = event["enabled"]
        elif "capsLock" in event:
            self.engine.set_caps_lock_on(event["capsLock"])
        elif "elevated" in event:
            self.platform.is_blocked_by_elevation = event["elevated"]
        elif "missedInput" in event:
            self.engine.reset_after_missed_input(event["missedInput"]["capsLock"])
        elif "config" in event:
            # The C# passes the vector's raw JSON text; re-serializing parses to the same config, and the parser's
            # own contract is covered by the 53 config vectors. Settled by the controller, not an oversight.
            config_text = json.dumps(event["config"])
            config = PARSER.parse(config_text).config
            if config is None:
                raise ValueError(f"Invalid config event: {config_text}")
            self.config = config
            self.engine.apply_config(config)
        else:
            raise ValueError(f"Unknown vector event: {json.dumps(event)}")

    def _press_chord(self, name: str, during: list[dict[str, Any]] | None) -> None:
        chord: Chord
        if name == "fixTyped":
            chord = self.config.fix_typed
        elif name == "fixSelection":
            chord = self.config.fix_selection
        else:
            raise ValueError(f"Unknown chord {name!r}.")
        holds = [_hold_key_name(hold) for hold in HOLD_KEY_ORDER if hold in chord.hold]
        trigger = NAMES.name_of(chord.trigger_vk)

        for hold in holds:
            self._key(hold, True)
        if during is not None:
            self.platform.during_action = lambda: self.run(during)
        self._key(trigger, True)
        self.platform.during_action = None
        self._key(trigger, False)
        for hold in reversed(holds):
            self._key(hold, False)

    def _key(self, name: str, down: bool) -> None:
        vk = NAMES.vk_of(name)
        if vk is None:
            raise ValueError(f"Unknown key name {name!r} in vector.")
        self._send(vk, down, is_packet=False)

    def _send(self, vk: int, down: bool, is_packet: bool) -> None:
        modifiers = HoldKeys.NONE
        for held in self._down:
            modifiers |= NAMES.hold_key_of(held)
        modifiers &= ~HoldKeys.CAPS_LOCK

        decision = self.engine.on_key(
            KeyEvent(vk, down, False, is_packet, modifiers, self.platform.window, self.platform.layout)
        )
        if down:
            self._down.add(vk)
        else:
            self._down.discard(vk)

        started = list(self._started)
        self._started.clear()

        text = "swallow" if decision.swallow else "pass"
        for action in started:
            text += " fire:fixTyped" if action is ChordAction.FIX_TYPED else " fire:fixSelection"
        for sent in decision.send_keys:
            text += " send:" + NAMES.name_of(sent)
        self.decisions.append(text)

        for action in started:
            self.engine.run_action(action)

File: csharp/tests/GibberishRewriter.Core.Tests/Harness/VectorRunner.cs
using System.Text.Json;

namespace GibberishRewriter.Core.Tests.Harness;

/// <summary>Loads a vector file and finds vectors by name. Names are unique within a file.</summary>
internal static class VectorFile
{
    public static IEnumerable<object[]> Names(string relative) =>
        Load(relative).Keys.Select(name => new object[] { name });

    public static JsonElement Get(string relative, string name) => Load(relative)[name];

    private static Dictionary<string, JsonElement> Load(string relative)
    {
        using var document = JsonDocument.Parse(SharedFiles.ReadText(relative));
        return document.RootElement.EnumerateArray()
            .ToDictionary(vector => vector.GetProperty("name").GetString()!, vector => vector.Clone());
    }
}

/// <summary>Feeds one vector's events to a real Engine with a FakePlatform (plan Task 10, vector harness rules).</summary>
internal sealed class VectorRunner
{
    public static readonly KeyNames Names = KeyNames.Load(SharedFiles.PathOf("keys.json"));
    public static readonly KeyTable Table = KeyTable.Load(SharedFiles.PathOf("layouts/us-arabic101.json"));

    private static readonly TextConverter Converter =
        new(Table, WordList.Load(SharedFiles.PathOf("wordlists/english.txt")));

    private static readonly ConfigParser Parser = new(Names, SharedFiles.ReadText("config.default.json"));

    private readonly HashSet<int> _down = [];
    private readonly Queue<ChordAction> _started = new();

    public VectorRunner(JsonElement vector)
    {
        var configText = vector.TryGetProperty("config", out var config) ? config.GetRawText() : "{}";
        Config = Parser.Parse(configText).Config
            ?? throw new InvalidOperationException($"Invalid config in vector: {configText}");

        Platform = new FakePlatform
        {
            Layout = vector.TryGetProperty("layout", out var layout) ? ParseLayout(layout.GetString()!) : LayoutKind.Latin,
            IsConsole = Flag(vector, "console"),
            IsBlockedByElevation = Flag(vector, "blockedByElevation"),
            ReleaseInTime = !Flag(vector, "releaseTimeout"),
            ClipboardText = vector.TryGetProperty("clipboard", out var clipboard) ? clipboard.GetString() : null,
            Selection = vector.TryGetProperty("selection", out var selection) ? selection.GetString() : null,
            ClipboardLocked = vector.TryGetProperty("clipboardLocked", out var locked) ? locked.GetString() : null,
        };

        Engine = new Engine(Names, Table, Converter, Config, Platform, _started.Enqueue, Flag(vector, "capsLock"));
    }

    public AppConfig Config { get; private set; }
    public FakePlatform Platform { get; }
    public Engine Engine { get; }

    /// <summary>One decision string per key event.</summary>
    public List<string> Decisions { get; } = [];

    /// <summary>Expected platform calls of a vector, as strings comparable with <see cref="FakePlatform.Calls"/>.</summary>
    public static List<string> ExpectedCalls(JsonElement vector) =>
        vector.GetProperty("expect").EnumerateArray()
            .Select(call => call.EnumerateObject().Single())
            .Select(field => field.Value.ValueKind == JsonValueKind.Number
                ? $"{field.Name} {field.Value.GetInt32()}"
                : $"{field.Name} {field.Value.GetString()}")
            .ToList();

    public void Run(JsonElement events)
    {
        foreach (var e in events.EnumerateArray())
        {
            Feed(e);
        }
    }

    private static bool Flag(JsonElement vector, string name) =>
        vector.TryGetProperty(name, out var value) && value.GetBoolean();

    private static LayoutKind? ParseLayout(string name) => name == "other" ? null : LayoutKinds.Parse(name);

    private static string HoldKeyName(HoldKeys hold) => hold switch
    {
        HoldKeys.Shift => "ShiftLeft",
        HoldKeys.Ctrl => "ControlLeft",
        HoldKeys.Alt => "AltLeft",
        HoldKeys.Win => "MetaLeft",
        _ => "CapsLock",
    };

    private void Feed(JsonElement e)
    {
        if (e.TryGetProperty("tap", out var tap))
        {
            var shift = Flag(e, "shift");
            if (shift)
            {
                Key("ShiftLeft", true);
            }
            Key(tap.GetString()!, true);
            Key(tap.GetString()!, false);
            if (shift)
            {
                Key("ShiftLeft", false);
            }
        }
        else if (e.TryGetProperty("down", out var down))
        {
            Key(down.GetString()!, true);
        }
        else if (e.TryGetProperty("up", out var up))
        {
            Key(up.GetString()!, false);
        }
        else if (e.TryGetProperty("chord", out var chordName))
        {
            PressChord(chordName.GetString()!, e.TryGetProperty("during", out var during) ? during.Clone() : null);
        }
        else if (e.TryGetProperty("mouseDown", out _))
        {
            Engine.OnMouseDown();
        }
        else if (e.TryGetProperty("window", out var window))
        {
            Platform.Window = window.GetInt64();
        }
        else if (e.TryGetProperty("layout", out var layout))
        {
            Platform.Layout = ParseLayout(layout.GetString()!);
        }
        else if (e.TryGetProperty("packet", out var packet))
        {
            foreach (var _ in packet.GetString()!)
            {
                Send(KeyNames.VkPacket, true, isPacket: true);
                Send(KeyNames.VkPacket, false, isPacket: true);
            }
        }
        else if (e.TryGetProperty("enabled", out var enabled))
        {
            Engine.Enabled = enabled.GetBoolean();
        }
        else if (e.TryGetProperty("capsLock", out var capsLock))
        {
            Engine.SetCapsLockOn(capsLock.GetBoolean());
        }
        else if (e.TryGetProperty("elevated", out var elevated))
        {
            Platform.IsBlockedByElevation = elevated.GetBoolean();
        }
        else if (e.TryGetProperty("missedInput", out var missed))
        {
            Engine.ResetAfterMissedInput(missed.GetProperty("capsLock").GetBoolean());
        }
        else if (e.TryGetProperty("config", out var reload))
        {
            Config = Parser.Parse(reload.GetRawText()).Config
                ?? throw new InvalidOperationException($"Invalid config event: {reload.GetRawText()}");
            Engine.ApplyConfig(Config);
        }
        else
        {
            throw new InvalidOperationException($"Unknown vector event: {e.GetRawText()}");
        }
    }

    private void PressChord(string name, JsonElement? during)
    {
        var chord = name switch
        {
            "fixTyped" => Config.FixTyped,
            "fixSelection" => Config.FixSelection,
            _ => throw new InvalidOperationException($"Unknown chord '{name}'."),
        };
        var holds = KeyNames.HoldKeyOrder.Where(hold => chord.Hold.HasFlag(hold)).Select(HoldKeyName).ToList();
        var trigger = Names.NameOf(chord.TriggerVk);

        foreach (var hold in holds)
        {
            Key(hold, true);
        }
        if (during is { } duringEvents)
        {
            Platform.DuringAction = () => Run(duringEvents);
        }
        Key(trigger, true);
        Platform.DuringAction = null;
        Key(trigger, false);
        for (var i = holds.Count - 1; i >= 0; i--)
        {
            Key(holds[i], false);
        }
    }

    private void Key(string name, bool down)
    {
        if (!Names.TryGetVk(name, out var vk))
        {
            throw new InvalidOperationException($"Unknown key name '{name}' in vector.");
        }
        Send(vk, down, isPacket: false);
    }

    private void Send(int vk, bool down, bool isPacket)
    {
        var modifiers = HoldKeys.None;
        foreach (var held in _down)
        {
            modifiers |= Names.HoldKeyOf(held);
        }
        modifiers &= ~HoldKeys.CapsLock;

        var decision = Engine.OnKey(new KeyEvent(vk, down, false, isPacket, modifiers, Platform.Window, Platform.Layout));
        if (down)
        {
            _down.Add(vk);
        }
        else
        {
            _down.Remove(vk);
        }

        var started = new List<ChordAction>();
        while (_started.TryDequeue(out var action))
        {
            started.Add(action);
        }

        var text = decision.Swallow ? "swallow" : "pass";
        foreach (var action in started)
        {
            text += action == ChordAction.FixTyped ? " fire:fixTyped" : " fire:fixSelection";
        }
        foreach (var sent in decision.SendKeys)
        {
            text += " send:" + Names.NameOf(sent);
        }
        Decisions.Add(text);

        foreach (var action in started)
        {
            Engine.RunAction(action);
        }
    }
}
